File: 2020/py/day_01.py
import itertools as it
from math import prod
from pathlib import Path


def find(iterable, r):
    for items in it.combinations(iterable, r):
        if sum(items) == 2020:
            return prod(items)


def main():
    text = (Path(__file__).parent / "../input/input_01.txt").read_text()
    entries = [int(line) for line in text.splitlines()]
    print("Part 1:", find(entries, 2))
    print("Part 2:", find(entries, 3))


if __name__ == "__main__":
    main()

File: 2020/rs/src/bin/day_01.rs
use std::fs;

fn find(vals: &[u64], sum: u64, prod: u64, depth: usize) -> Option<u64> {
    for (i, &x) in vals.iter().enumerate() {
        let new_sum = sum + x;
        let new_prod = prod * x;
        if new_sum > 2020 {
            return None;
        }
        if depth == 1 {
            if new_sum == 2020 {
                return Some(new_prod);
            }
        } else if let Some(x) = find(&vals[i+1..], new_sum, new_prod, depth-1) {
            return Some(x);
        }
    }
    None
}

fn main() {
    let input = fs::read_to_string("../input/input_01.txt").unwrap();
    let mut entries: Vec<u64> = input.split_terminator('\n')
        .map(|s| s.parse().unwrap())
        .collect();
    entries.sort();
    println!("Part 1: {}", find(&entries, 0, 1, 2).unwrap());
    println!("Part 2: {}", find(&entries, 0, 1, 3).unwrap());
}
